File: ios/FrenchNCLC7/Models/DayPlan.swift
import Foundation

/// One section of the day-plan PDF: a title, and either flashcards, a paragraph, or a list.
struct PlanPdfSection: Equatable {
    let title: String
    /// Small text on the right of the heading, e.g. "12 new cards".
    var note: String? = nil
    var cards: [AnkiCard] = []
    var text: String? = nil
    var items: [String] = []
}

/// Everything that goes in the PDF for one day (kept apart from drawing, so it can be tested).
struct DayPlanContent: Equatable {
    let day: Int
    let week: Int
    let sections: [PlanPdfSection]
}

/// The download allowance: whether another PDF may be downloaded now, and when the next slot frees up.
struct PdfQuota: Equatable {
    let allowed: Bool
    let resetsAt: String?

    func resetDate() -> Date? {
        guard let text = resetsAt else { return nil }
        // The server sends microseconds ("...53.064937+00:00"); read the whole seconds, then add the fraction back.
        var whole = text
        var fraction = 0.0
        if let dot = text.firstIndex(of: "."),
           let end = text[dot...].firstIndex(where: { $0 == "+" || $0 == "-" || $0 == "Z" }) {
            fraction = Double("0" + String(text[dot..<end])) ?? 0
            whole = String(text[..<dot]) + String(text[end...])
        }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: whole).map { $0.addingTimeInterval(fraction) }
    }

    func secondsUntilReset(now: Date = Date()) -> TimeInterval {
        (resetDate() ?? now).timeIntervalSince(now)
    }
}

/// The reply to asking for a download: `ok` is true when a download was reserved (`id` is the reservation).
struct PdfClaim: Equatable {
    let ok: Bool
    let id: Int?
    let status: PdfQuota?
}

enum DayPlanLogic {
    private static func splitRaw(_ text: String) -> [String] {
        text.split(separator: ";", omittingEmptySubsequences: true)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    /// The day's plan across the five sections, as plain text (no links), matching the web app's PDF:
    /// Anki cards, the grammar task, Kwiziq lessons, TV5MONDE (its level, then the lessons), and the writing task.
    static func build(anki: DayContent, grammar: DayContent, kwiziq: DayContent, tv5: DayContent, writing: DayContent) -> DayPlanContent {
        DayPlanContent(day: anki.day, week: anki.week, sections: [
            PlanPdfSection(title: "Anki", note: "\(anki.cards.count) new cards", cards: anki.cards),
            PlanPdfSection(title: "Grammar book", text: grammar.text),
            PlanPdfSection(title: "Kwiziq", items: splitRaw(kwiziq.text)),
            PlanPdfSection(title: "TV5MONDE", items: (tv5.badge.map { [$0] } ?? []) + splitRaw(tv5.text)),
            PlanPdfSection(title: "Writing", text: writing.text),
        ])
    }

    static func fileName(_ day: Int) -> String { "French-NCLC7-Day-\(day).pdf" }

    /// "Available again in 5h 12m"
    static func waitText(_ quota: PdfQuota, now: Date = Date()) -> String {
        "Available again in " + WritingLogic.formatWait(quota.secondsUntilReset(now: now))
    }

    private static func quota(from value: Any?) -> PdfQuota? {
        guard let obj = value as? [String: Any], let allowed = obj["allowed"] as? Bool else { return nil }
        return PdfQuota(allowed: allowed, resetsAt: obj["resets_at"] as? String)
    }

    /// Reads the reply to "where do I stand" (nil if it can't be read).
    static func parseQuota(_ body: Data) -> PdfQuota? {
        quota(from: try? JSONSerialization.jsonObject(with: body))
    }

    /// Reads the reply to "reserve a download". An unreadable reply counts as not allowed (never as allowed).
    static func parseClaim(_ body: Data) -> PdfClaim {
        guard let obj = (try? JSONSerialization.jsonObject(with: body)) as? [String: Any] else {
            return PdfClaim(ok: false, id: nil, status: nil)
        }
        let ok = (obj["ok"] as? Bool) == true
        return PdfClaim(ok: ok, id: (obj["id"] as? NSNumber)?.intValue, status: quota(from: obj["status"]))
    }
}

File: android/app/src/main/java/com/frenchnclc7/app/data/DayPlan.kt
package com.frenchnclc7.app.data

import java.time.Instant
import java.time.OffsetDateTime
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** One section of the day-plan PDF: a title, and either flashcards, a paragraph, or a list. */
data class PlanPdfSection(
    val title: String,
    /** Small text on the right of the heading, e.g. "12 new cards". */
    val note: String? = null,
    val cards: List<AnkiCard> = emptyList(),
    val text: String? = null,
    val items: List<String> = emptyList(),
)

/** Everything that goes in the PDF for one day (kept apart from drawing, so it can be tested). */
data class DayPlanContent(val day: Int, val week: Int, val sections: List<PlanPdfSection>)

/** The download allowance: whether another PDF may be downloaded now, and when the next slot frees up. */
data class PdfQuota(val allowed: Boolean, val resetsAt: String?) {
    fun resetInstant(): Instant? = resetsAt?.let { try { OffsetDateTime.parse(it).toInstant() } catch (e: Exception) { null } }

    fun millisUntilReset(now: Instant = Instant.now()): Long = (resetInstant()?.toEpochMilli() ?: now.toEpochMilli()) - now.toEpochMilli()
}

/** The reply to asking for a download: [ok] is true when a download was reserved ([id] is the reservation). */
data class PdfClaim(val ok: Boolean, val id: Long?, val status: PdfQuota?)

object DayPlanLogic {
    private fun splitRaw(text: String) = text.split(";").map { it.trim() }.filter { it.isNotEmpty() }

    /**
     * The day's plan across the five sections, as plain text (no links), matching the web app's PDF:
     * Anki cards, the grammar task, Kwiziq lessons, TV5MONDE (its level, then the lessons), and the writing task.
     */
    fun build(anki: DayContent, grammar: DayContent, kwiziq: DayContent, tv5: DayContent, writing: DayContent): DayPlanContent =
        DayPlanContent(
            day = anki.day,
            week = anki.week,
            sections = listOf(
                PlanPdfSection("Anki", note = anki.cards.size.toString() + " new cards", cards = anki.cards),
                PlanPdfSection("Grammar book", text = grammar.text),
                PlanPdfSection("Kwiziq", items = splitRaw(kwiziq.text)),
                PlanPdfSection("TV5MONDE", items = listOfNotNull(tv5.badge) + splitRaw(tv5.text)),
                PlanPdfSection("Writing", text = writing.text),
            ),
        )

    fun fileName(day: Int): String = "French-NCLC7-Day-$day.pdf"

    /** "Available again in 5h 12m" */
    fun waitText(quota: PdfQuota, now: Instant = Instant.now()): String =
        "Available again in " + WritingLogic.formatWait(quota.millisUntilReset(now))

    private fun quotaFrom(el: JsonElement?): PdfQuota? {
        val obj = el as? JsonObject ?: return null
        val allowed = (obj["allowed"] as? JsonPrimitive)?.contentOrNull?.toBooleanStrictOrNull() ?: return null
        val resets = obj["resets_at"]?.let { if (it is JsonPrimitive && it.isString) it.content else null }
        return PdfQuota(allowed, resets)
    }

    /** Reads the reply to "where do I stand" (null if it can't be read). */
    fun parseQuota(body: String): PdfQuota? = try { quotaFrom(Json.parseToJsonElement(body)) } catch (e: Exception) { null }

    /** Reads the reply to "reserve a download". An unreadable reply counts as not allowed (never as allowed). */
    fun parseClaim(body: String): PdfClaim = try {
        val obj = Json.parseToJsonElement(body).jsonObject
        val ok = obj["ok"]?.jsonPrimitive?.contentOrNull?.toBooleanStrictOrNull() == true
        PdfClaim(ok, obj["id"]?.jsonPrimitive?.contentOrNull?.toLongOrNull(), quotaFrom(obj["status"]))
    } catch (e: Exception) {
        PdfClaim(false, null, null)
    }
}
